python/cuml/common/array_sparse.py

```diff
@@ -0,0 +1,171 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from cuml.common.import_utils import has_scipy
+
+import cupyx as cpx
+import numpy as np
+
+from cuml.common.memory_utils import with_cupy_rmm
+
+from cuml.common.input_utils import input_to_cuml_array
+
+if has_scipy():
+    import scipy.sparse
+
+
+class SparseCumlArray:
+    """
+    SparseCumlArray abstracts sparse array allocations. This will
+    accept either a Scipy or Cupy sparse array and construct CumlArrays
+    out of the underlying index and data arrays. Currently, this class
+    only supports the CSR array format and input in any other sparse
+    format will be converted to CSR.
+
+    Parameters
+    ----------
+
+    data : scipy.sparse.spmatrix or cupyx.scipy.sparse.spmatrix
+        A Scipy or Cupy sparse matrix
+    convert_to_dtype : data-type or False, optional
+        Any object that can be interpreted as a numpy or cupy data type.
+        Specifies whether to convert the data array to a different dtype.
+    convert_index : data-type or False (default: np.int32), optional
+        Any object that can be interpreted as a numpy or cupy data type.
+        Specifies whether to convert the indices to a different dtype. By
+        default, it is preferred to use 32-bit indexing.
+    convert_format : bool, optional (default: False)
```

Comment:
I like that this parameter is here, I was wondering if we could open an issue to revisit whether False or True should be the default? One of the things I learnt with `convert_dtype` and similar is that `True` ended up being the better default for users so far, so I think it would be good to revisit the default after having this in the wild for a version or two.

Reply:
I don't mind setting this to true by default. Really, I was thinking about the cases John brought up where users could get a surprise if they pass in a CSR and no memory is copied but a non-CSR format will cause a copy. What are your thoughts about issuing a warning in that scenario, similarly to CumlArray when the order needs to be changed?

Reply:
Yeah, there's a fair chance we'll end up wanting to change it... I feel like False is a bit more conservative about surprising the user, so it is a good start though.